fields.tex

```diff
@@ -3,6 +3,79 @@ \chapter{Fields and vector spaces}
 
 Quotients; subspaces (= ?). Bases and so. Dual space; orthogonality. (all of this depends on good implementations of subobjects). Eigen-stuff. Characteristic polynomials; Hamilton-Cayley.
 \section{the algebraic hierarchy: groups, abelian groups, rings, fields}
+
+\begin{definition}
+Given a commutative ring $R$, an element $e:R$ is \textbf{invertible} if there exists an element $a:R$ such that $e \cdot a = 1$ and $a \cdot e = 1$:
+$$\mathrm{isInvertible}(e) := \left\Vert\sum_{a:R} (e \cdot a = 1) \times (a \cdot e = 1)\right\Vert$$
+\end{definition}
+
+\begin{theorem}
+One is always an invertible element of $R$. 
+\end{theorem}
+
+\begin{proof}
+$1 \cdot 1 = 1$. Thus, there exists an element $1:R$ such that $1 \cdot 1 = 1$ and $1 \cdot 1 = 1$, and $1:R$ is invertible. 
+\end{proof}
+
+The following definition of non-invertible element is due to Henri Lombardi and Claude Quitté in \textit{Commutative algebra: Constructive methods (Finite projective modules)}: 
+\begin{definition}
+An element $x:R$ is \textbf{non-invertible} if $x$ being invertible implies that $0 = 1$
+$$\mathrm{isNonInvertible}(x) := \mathrm{isInvertible}(x) \to (0 = 1)$$
+\end{definition}
+
+\begin{theorem}
+Zero is always a non-invertible element of $R$. 
+\end{theorem}
+
+\begin{proof}
+Suppose that $0$ is invertible. Then there exists an element $a:R$ such that $a \cdot 0 = 1$. However, due to the absorption properties of $0$ and the fact that $R$ is a set, $a \cdot 0 = 0$. This implies that $0 = 1$. Since $0$ being invertible implies $0 = 1$, $0$ is a non-invertible element of $R$. 
+\end{proof}
+
+As a result, the type of all non-invertible elements in $R$ is inhabited. 
+
+\begin{theorem}
+If $R$ is the trivial ring, then $0$ is both invertible and non-invertible.
+\end{theorem}
+
+\begin{proof}
+$1$ is always invertible, and $0$ is always non-invertible. Since in a trivial ring, $0 = 1$, and $R$ is a set, $0$ is always invertible as well. 
+\end{proof}
+
+\begin{definition}
+A commutative ring $R$ is a \textbf{field} if and only if the type of all non-invertible elements in $R$ is contractible:
+$$\mathrm{isField}(R) := \mathrm{isCRing}(R) \times \mathrm{isContr}\left(\sum_{x:R} \mathrm{isNonInvertible}(x)\right)$$ 
+Equivalently, $R$ is a field if and only if every non-invertible element is equal to zero. 
+\end{definition}
+
+\begin{remark}
+In other parts of the constructive mathematics literature, such as in Peter Johnstone's \textit{Rings, Fields, and Spectra}, this is called a "residue field". However, in this book we shall refrain from using the term "residue field" for our definition, since that contradicts the usage of "residue field" in other parts of mathematics, such as in algebraic geometry. 
+\end{remark}
+
+\begin{remark}
+Most authors additionally assert that every field is a non-trivial ring ($0 \neq 1$). The authors of this book have decided to follow the approach of Lombardi and Quitté, where the trivial ring is a field. 
```

Comment:
Not making the trivial ring into a field might be better.  For example, one might like to say that the quotient rings of R that are fields correspond to the maximal ideals.  Or one might like to say that the dimension of a vector space over a field is a well-defined invariant.